src/main/java/org/spongepowered/common/mixin/core/advancement/MixinAdvancement.java

```diff
@@ -89,10 +92,22 @@
 
     @Nullable private Advancement tempParent;
 
+    // Whether the Advancement is constructed for the client
+    private boolean client;
+
+    private void checkServer() {
+        checkState(!this.client);
+    }
+
     @SuppressWarnings("ConstantConditions")
     @Inject(method = "<init>", at = @At("RETURN"))
     private void onInit(ResourceLocation id, @Nullable Advancement parentIn, @Nullable DisplayInfo displayIn,
             AdvancementRewards rewardsIn, Map<String, Criterion> criteriaIn, String[][] requirementsIn, CallbackInfo ci) {
+        this.client = SpongeAdvancementHelper.CONSTRUCTING_CLIENT_ADVANCEMENTS.get();
```

Comment:
No..... Should be able to rely on other methods than just relying on thread local booleans.